src/Pkg/Package.hs

```diff
@@ -287,20 +286,24 @@ buildMods opts ns = do let f = map (toPath . showCG) ns
     where toPath n = foldl1' (</>) $ splitOn "." n
 
 testLib :: Bool -> String -> String -> IO Bool
-testLib warn p f
-    = do d <- getDataDir
-         gcc <- getCC
-         (tmpf, tmph) <- tempfile ""
-         hClose tmph
-         let libtest = d </> "rts" </> "libtest.c"
-         e <- rawSystem gcc [libtest, "-l" ++ f, "-o", tmpf]
-         case e of
-            ExitSuccess -> return True
-            _ -> do if warn
-                       then do putStrLn $ "Not building " ++ p ++
-                                          " due to missing library " ++ f
-                               return False
-                       else fail $ "Missing library " ++ f
+testLib warn p f = do
+    d <- getDataDir
+    gcc <- getCC
+    (tmpf, tmph) <- tempfile ""
+    hClose tmph
+    let libtest = d </> "rts" </> "libtest.c"
+    e <- rawSystem gcc [libtest, "-l" ++ f, "-o", tmpf]
+    case e of
+       ExitSuccess -> return True
+       _ -> if warn
+              then do
+                putStrLn $ unwords ["Not building"
+                                   , p
+                                   , "due to missing library"
+                                   , f
+                                   ]
+                return False
```

Comment:
Here's an example where I think dangling "do" mode really shows its weakness. There's no "<-" binds, so there's no visual cue that we're in a "do", and scanning up the margin doesn't reveal it.